plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,253 @@
+
+    //
+    // Function Implementations
+    //
+
+    auto call_switch_user(irods::api_entry* _api, RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        return _api->call_handler<SwitchUserInput*>(_comm, _input);
+    } // call_switch_user
+
+    auto check_input(const SwitchUserInput* _input) -> int
+    {
+        if (!_input) {
+            log_api::error("Invalid input argument: received a null pointer");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->username) == 0) {
+            log_api::error("Invalid username argument: empty string");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        if (std::strlen(_input->zone) == 0) {
+            log_api::error("Invalid zone argument: empty string");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        return 0;
+    } // check_input
+
+    auto update_server_to_server_connections(RsComm& _comm) -> void
+    {
+        const auto switch_user_or_disconnect = [&client = _comm.clientUser](rodsServerHost& _host) {
+            if (!_host.conn) {
+                log_api::trace("No connection to remote host [{}].", _host.hostName->name);
+                return;
+            }
+
+            // Disconnect if the remote server's version is older than 4.3.1.
+            // Remember, only iRODS 4.3.1 and later support the switch user API plugin.
+            if (*irods::to_version(_host.conn->svrVersion->relVersion) < irods::version{4, 3, 1}) {
+                rcDisconnect(_host.conn);
+                _host.conn = nullptr;
+                return;
+            }
+
+            // At this point, we know the remote server supports this API plugin.
+            if (const auto ec = rc_switch_user(_host.conn, client.userName, client.rodsZone); ec != 0) {
+                log_api::error(
+                    "rc_switch_user failed on remote host [{}] with error_code [{}]. Disconnecting from remote host.",
+                    _host.hostName->name,
+                    ec);
+                rcDisconnect(_host.conn);
+                _host.conn = nullptr;
+            }
+        };
+
+        for (auto* zone_ptr = ZoneInfoHead; zone_ptr; zone_ptr = zone_ptr->next) {
+            for (auto* host_ptr = zone_ptr->primaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+
+            for (auto* host_ptr = zone_ptr->secondaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+        }
+    } // update_server_to_server_connections
+
+    auto rs_switch_user(RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        try {
+            // Only administrators are allowed to invoke this API.
+            // We check the proxy user because it covers clients and server-to-server redirects.
+            // Remember, the client and proxy user information is identical until there's a redirect.
```

Comment:
Is this always true?

Reply:
Yes, unless you've seen something different.

Reply:
Okay. I may be getting authentication and connection information conflated in my mental model.

Reply:
I believe there is a comment in the server that explicitly states this. Not sure where I found that. I know rcConnect sets both to the same value depending on how it is invoked.